Cap the duration of EventLoop::run in the main thread

rniwa opened the pull request:
#### 9c588230925217e6bd22cb40faaba28dac210e1a
<pre>
Cap the duration of EventLoop::run in the main thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=260112">https://bugs.webkit.org/show_bug.cgi?id=260112</a>

Reviewed by Yusuke Suzuki.

Cap the duration of EventLoop::run in the main thread using the same time limit
ThreadTimers::sharedTimerFiredInternal() uses.

* LayoutTests/fast/eventloop/queue-task-event-loop-starvation-expected.txt: Added.
* LayoutTests/fast/eventloop/queue-task-event-loop-starvation.html: Added.
* Source/WebCore/dom/EventLoop.cpp:
(WebCore::EventLoop::run):
* Source/WebCore/dom/EventLoop.h:
* Source/WebCore/dom/WindowEventLoop.cpp:
(WebCore::WindowEventLoop::didReachTimeToRun):
* Source/WebCore/platform/ThreadTimers.cpp:
* Source/WebCore/platform/ThreadTimers.h:

Canonical link: <a href="https://commits.webkit.org/266896@main">https://commits.webkit.org/266896@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6500039caf2c92698d332f944cee511fff280536

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/15053 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15358 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15710 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16807 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14154 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/15191 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17874 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15457 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/16782 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15234 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15698 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12788 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17536 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12978 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13574 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/20541 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14056 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13742 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16985 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14305 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/12106 "11 flakes 1 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13586 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/13431 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17923 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1817 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14146 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/WebCore/dom/EventLoop.h:
```diff
@@ -121,7 +123,7 @@ class EventLoop : public RefCounted<EventLoop>, public CanMakeWeakPtr<EventLoop>
 protected:
     EventLoop();
-    void run();
+    void run(Markable<ApproximateTime> deadline = std::nullopt);
```
Markable is used when it is member (for size compaction) There are automatic conversion so use std::optional<ApproximateTime> here.

rniwa replied:
Sure, will fix.

Constellation reviewed (approved):
r=me

rniwa committed:
Cap the duration of EventLoop::run in the main thread
https://bugs.webkit.org/show_bug.cgi?id=260112

Reviewed by Yusuke Suzuki.

Cap the duration of EventLoop::run in the main thread using the same time limit
ThreadTimers::sharedTimerFiredInternal() uses.

* LayoutTests/fast/eventloop/queue-task-event-loop-starvation-expected.txt: Added.
* LayoutTests/fast/eventloop/queue-task-event-loop-starvation.html: Added.
* Source/WebCore/dom/EventLoop.cpp:
(WebCore::EventLoop::run):
* Source/WebCore/dom/EventLoop.h:
* Source/WebCore/dom/WindowEventLoop.cpp:
(WebCore::WindowEventLoop::didReachTimeToRun):
* Source/WebCore/platform/ThreadTimers.cpp:
* Source/WebCore/platform/ThreadTimers.h:

Canonical link: https://commits.webkit.org/266896@main
